server/routers/wallets/createPublicWallet.ts

```diff
@@ -59,42 +61,53 @@ export const CreatePublicWalletInputSchema = z
 export const createPublicWallet = protectedProcedure
   .input(CreatePublicWalletInputSchema)
   .mutation(async ({ input, ctx }) => {
-    const canBeRecovered =
-      input.source.type === WalletSourceType.IMPORTED ? true : false;
+    try {
+      // First get device ID
+      const deviceId = await getDeviceAndLocationId(ctx);
 
-    const wallet = await ctx.prisma.wallet.create({
-      data: {
+      // Create wallet using the security definer function
+      const walletId = await createWalletWithSecurityDefiner(ctx, {
         status: input.status,
         chain: input.chain,
         address: input.address,
         publicKey: input.publicKey,
+        identifierTypeSetting: WalletIdentifierType.ALIAS,
         aliasSetting: input.aliasSetting,
-        descriptionSetting: input.descriptionSetting,
-        tagsSetting: input.tagsSetting,
-        doNotAskAgainSetting: false,
-        walletPrivacySetting: WalletPrivacySetting.PUBLIC,
-        canRecoverAccountSetting: input.canRecoverAccountSetting,
-        canBeRecovered,
-        source: input.source as InputJsonValue,
+        deviceAndLocationId: deviceId,
+      });
 
-        userProfile: getUserConnectOrCreate(ctx),
-
-        deviceAndLocation: getDeviceAndLocationConnectOrCreate(ctx),
-
-        workKeyShares: {
-          create: {
-            authShare: input.authShare,
-            deviceShareHash: input.deviceShareHash,
-            deviceSharePublicKey: input.deviceSharePublicKey,
-            userId: ctx.user.id,
-            sessionId: ctx.session.id,
-            deviceNonce: ctx.session.deviceNonce,
+      // Add the additional data after creating the wallet
+      const canBeRecovered = input.source.type === WalletSourceType.IMPORTED ? true : false;
+      
+      // Update the wallet with the additional fields
+      const wallet = await ctx.prisma.wallet.update({
+        where: { id: walletId },
```

Comment:
It's s good practice to still include the `userId`, as if we ever have some kind of issue with RLS, we'll still avoid returning data that belongs to other users.